src/react/hooks/useFragment.ts

```diff
@@ -46,48 +47,57 @@ export function useFragment<TData = any, TVars = OperationVariables>(
 ): UseFragmentResult<TData> {
   const { cache } = useApolloClient();
 
-  const { fragment, fragmentName, from, optimistic = true, ...rest } = options;
+  const diffOptions = useDeepMemo<Cache.DiffOptions<TData, TVars>>(() => {
+    const {
+      fragment,
+      fragmentName,
+      from,
+      optimistic = true,
+      ...rest
+    } = options;
 
-  const diffOptions: Cache.DiffOptions<TData, TVars> = {
-    ...rest,
-    returnPartialData: true,
-    id: typeof from === "string" ? from : cache.identify(from),
-    query: cache["getFragmentDoc"](fragment, fragmentName),
-    optimistic,
-  };
+    return {
+      ...rest,
+      returnPartialData: true,
+      id: typeof from === "string" ? from : cache.identify(from),
+      query: cache["getFragmentDoc"](fragment, fragmentName),
+      optimistic,
+    };
+  }, [options]);
 
-  const resultRef = React.useRef<UseFragmentResult<TData>>();
-  let latestDiff = cache.diff<TData>(diffOptions);
+  const resultRef = useLazyRef<UseFragmentResult<TData>>(() =>
+    diffToResult(cache.diff<TData>(diffOptions))
+  );
 
   // Used for both getSnapshot and getServerSnapshot
-  const getSnapshot = () => {
-    const latestDiffToResult = diffToResult(latestDiff);
-    return (
-        resultRef.current &&
-          equal(resultRef.current.data, latestDiffToResult.data)
-      ) ?
-        resultRef.current
-      : (resultRef.current = latestDiffToResult);
-  };
+  const getSnapshot = React.useCallback(() => resultRef.current, []);
 
   return useSyncExternalStore(
-    (forceUpdate) => {
-      let lastTimeout = 0;
-      const unsubcribe = cache.watch({
-        ...diffOptions,
-        immediate: true,
-        callback(diff) {
-          if (!equal(diff, latestDiff)) {
-            resultRef.current = diffToResult((latestDiff = diff));
-            lastTimeout = setTimeout(forceUpdate) as any;
-          }
-        },
-      });
-      return () => {
-        unsubcribe();
-        clearTimeout(lastTimeout);
-      };
-    },
+    React.useCallback(
+      (forceUpdate) => {
+        let lastTimeout = 0;
+        const unsubcribe = cache.watch({
```

Comment:
```suggestion
        const unsubscribe = cache.watch({
```

While we're at it, let's fix this typo 😅

Reply:
ha, good catch! [58384c3](https://github.com/apollographql/apollo-client/pull/11464/commits/58384c322fd065bd8b178f0c4d025aefca20099c)